### ISSUE TITLE
sizeof on enums

### ISSUE THREAD

**Issue by samphunter** [Southclaws/pawn#237]
This compiler changes the way sizeof handles enums when used as default parameter. Wouldn't it be better if it behaved in the same way when used as standalone?
Example:
```C
enum HOUSE_INFO {
    HouseName[32],
    HousePrice,
    HouseLook
}
new HouseData[HOUSE_INFO];

SetName(const name[]) {
    format(HouseData[HouseName], sizeof(HouseName), name);
}
```
It would be more intuitive for sizeof() to act the same regardless of where it is used and using sizeof on enum item right now is pointless anyway, as it always returns 1. So this should not break anything, unless I am missing something.

PS: also sizeof(HOUSE_INFO) would be nice.

**Comment by YashasSamaga** [comment]
What do you mean by `sizeof(HOUSE_INFO)`? `HOUSE_INFO` already stores the size which is why enum arrays work in the first place. It works because `HOUSE_INFO` is replaced with the actual size of the enum array and the constants declared inside the `enum` would be given values which access the correct index of the array.

Try `printf("%d", HOUSE_INFO);`.

**Comment by Y-Less** [comment]
I've explained why enums are not arrays SO MANY times, but apparently I need to do it again.

This:

```pawn
enum HOUSE_INFO {
    HouseName[32],
    HousePrice,
    HouseLook
}
new HouseData[HOUSE_INFO];
```

Is syntactic sugar for this:

```pawn
enum HOUSE_INFO {
    HouseName,
    HousePrice = HouseName + 32,
    HouseLook
}
new HouseData[HOUSE_INFO];
```

If you can't see why `sizeof (HouseName)` is `1` from that, let me go futher.  That is syntactic sugar for this:

```pawn
const HOUSE_INFO:HOUSE_INFO = HOUSE_INFO:34;
const HOUSE_INFO:HouseName = HOUSE_INFO:0;
const HOUSE_INFO:HousePrice = HOUSE_INFO:32;
const HOUSE_INFO:HouseLook = HOUSE_INFO:33;

new HouseData[HOUSE_INFO];
```

So `sizeof (HouseName)` is undeniably `1`.  People just like to believe that enums and arrays are the same thing - they are not.

**Comment by samphunter** [comment]
No, I understand enums are not arrays. But we already changed one of the sizeof operators to return the "size" anyway and I just don't see the point of not changing the other one as well. Like in what case would getting the size of enum be ever useful? It is always 1.

Tho clarify, this is not a bug report, it is a suggestion.

**Comment by Zeex** [comment]
I think it could break a lot of code if we modified how `sizeof` works, even if we made it better. Maybe we should make a new operator that can return size of enum fields? `sizeoffield`?

**Comment by samphunter** [comment]
It could be, and if there is any code using sizeof in this way, then I completely agree. However, what code would be using it? Why would anyone need the sizeof of a constant, which can not be anything other than 1?

The solution with a warning and a new operator of course have the problem, that libraries written for the original compiler would have to be rewritten to use this new operator. That would include functions in the default includes, such as strcat.

And you could of course disable this behavior with the compatibility switch, so I don't really think it is such a big problem.

Of course, correct me if I am wrong in my assumptions.

**Comment by Zeex** [comment]
I completely forgot that there is compatibility mode. In this I think it's better to leave the old behavior of `sizeof` in compatibility mode and make it better in normal mode without inventing a new name.

**Comment by IllidanS4** [comment]
The default parameter is resolved to `sizeof(HouseData[HouseName])` which returns 32, as expected. Also `sizeof` is not the only operator to be considered, since `tagof` on enum fields also returns the tag of the enum itself and not the "inner" tag, whatever you call it.

Also the example gives me "constant symbol has no size" on the old compiler. That sounds more reasonable.
